### ISSUE TITLE
Regression: Whitespace problems again

### ISSUE THREAD

**Issue by l4t3b0** [pfidr34/docker-rclone#21]
Hi pfidr34

Sorry i did not react on issue #16. i was busy nowdays.

I just checked out your latest modification and i would like to report the same issue like #13.
Now i have problems again with the whitespace character :(

This is the result of the rclone command for the same config that was working already:

Command sync needs 2 arguments maximum: you provided 3 non flag arguments: ["webdav-alapitvany-private:/Megosztott" "dokumentumok/" "/data"]

If you remember there is a space between "webdav-alapitvany-private:/Megosztott" and "dokumentumok" but it should be one string.

**Comment by pfidr34** [comment]
Sorry, stepped on the change in the last update. New commit added and new builds up on Docker Hub.

**Comment by l4t3b0** [comment]
Hi pfidr34

Thank you for the quick response

I have just tried it, but something is not ok yet.
I still have the issue with your update:
```
2020-06-09 17:28:12 (362 KB/s) - '/dev/null' saved [2/2]

INFO: Removing logs older than 120 day(s)...
INFO: Source directory is not empty and can be processed without clear loss of data
INFO: Log file output to /logs/2020_06_09-17_28_14.txt
INFO: Starting rclone sync webdav-alapitvany-private:/Megosztott dokumentumok/ /data --config /config/rclone.conf -v --exclude /Forms/** --log-file=/logs/2020_06_09-17_28_14.txt
INFO: Sending failure signal to healthchecks.io
--2020-06-09 17:28:14-- https://hc-ping.com/d34f5437-......-9545b3f023cd/fail
Resolving hc-ping.com... 176.9.71.146, 178.63.26.145, 159.69.66.229, ...
Connecting to hc-ping.com|176.9.71.146|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 2 [text/plain]
Saving to: '/dev/null'
```
From the rclonde log:
`Command sync needs 2 arguments maximum: you provided 3 non flag arguments: ["webdav-alapitvany-private:/Megosztott" "dokumentumok/" "/data"]`

I can confirm that the update was successful. Here is a snippet from my local sync.sh:
```
    if [ ! -z "$OUTPUT_LOG" ]
    then
      d=$(date +%Y_%m_%d-%H_%M_%S)
      LOG_FILE="/logs/$d.txt"
      echo "INFO: Log file output to $LOG_FILE"
      echo "INFO: Starting rclone $RCLONE_CMD $SYNC_SRC $SYNC_DEST $RCLONE_OPTS $SYNC_OPTS_ALL --log-file=${LOG_FILE}"
      set +e
      eval "rclone $RCLONE_CMD "$(eval echo $SYNC_SRC)" "$(eval echo $SYNC_DEST)" $RCLONE_OPTS $SYNC_OPTS_ALL --log-file=${LOG_FILE}"
      export RETURN_CODE=$?
      set -e
    else
      echo "INFO: Starting rclone $RCLONE_CMD $SYNC_SRC $SYNC_DEST $RCLONE_OPTS $SYNC_OPTS_ALL"
      set +e
      eval "rclone $RCLONE_CMD "$(eval echo $SYNC_SRC)" "$(eval echo $SYNC_DEST)" $RCLONE_OPTS $SYNC_OPTS_ALL"
      set -e
      export RETURN_CODE=$?
    fi
    else
      echo "WARNING: Source directory is empty. Skipping $RCLONE_CMD command."
    fi
```

Maybe it has something to do with the embedded double eval expression?!?!

**Comment by pfidr34** [comment]
I rolled back the change. I tested last night and it is working. 

Make sure your paths are wrapped in single ticks.

Docker Compose (notice spaces in `SYNC_SRC` and `SYNC_DEST`):
  ```
rclone_test:
    image: "rclone_test"
    container_name: "rclone_test"
    volumes:
      - ${USERDIR}/docker/rclone_test:/config
      - ${USERDIR}/docker/rclone_test/logs:/logs
      - /mnt/SSD_STRGE/TEMP/:/mnt/TEMP
    environment:
      - UID=1000
      - GID=1000
      - CRON=@daily
      - RCLONE_CMD=copy
      - SYNC_SRC='/mnt/TEMP/TEST DIR'
      - SYNC_DEST='GDrive:/TEST DEST'
      - SYNC_OPTS=--fast-list --transfers 10 -v --bwlimit "06:00,200 00:00,off"
      - RCLONE_DIR_CMD=lsf
      - RCLONE_DIR_CMD_DEPTH=1
      - OUTPUT_LOG=1
      - ROTATE_LOG=7
      - CHECK_URL=https://hc-ping.com/<snip>
      - FORCE_SYNC=1
      - TZ=${TZ}
```

Container log:
```
INFO: Running rclone v1.52.0,
INFO: Cron shortcut @daily re-written to 0 0 * * *,
INFO: Starting sync.sh pid 45 Tue Jun  9 18:03:55 CDT 2020,
INFO: Sending start signal to healthchecks.io,
--2020-06-09 18:03:55--  https://hc-ping.com/<snip>/start,
Resolving hc-ping.com... 178.63.26.145, 176.9.71.146, 159.69.66.229, ...,
Connecting to hc-ping.com|178.63.26.145|:443... connected.,
HTTP request sent, awaiting response... 200 OK,
Length: 2 [text/plain],
Saving to: '/dev/null',
,
     0K                                                       100% 73.9K=0s,
,
2020-06-09 18:03:55 (73.9 KB/s) - '/dev/null' saved [2/2],
,
INFO: Removing logs older than 7 day(s)...,
INFO: Source directory is not empty and can be processed without clear loss of data,
INFO: Log file output to /logs/2020_06_09-18_03_55.txt,
INFO: Starting rclone copy '/mnt/TEMP/TEST DIR' 'GDrive:/TEST DEST' --config /config/rclone.conf --fast-list --transfers 10 -v --bwlimit "06:00,200 00:00,off" --log-file=/logs/2020_06_09-18_03_55.txt,
INFO: Sending complete signal to healthchecks.io,
--2020-06-09 18:03:58--  https://hc-ping.com/<snip>,
Resolving hc-ping.com... 176.9.71.146, 178.63.26.145, 159.69.66.229, ...,
Connecting to hc-ping.com|176.9.71.146|:443... connected.,
HTTP request sent, awaiting response... 200 OK,
Length: 2 [text/plain],
Saving to: '/dev/null',
,
     0K                                                       100%  114K=0s,
,
2020-06-09 18:03:59 (114 KB/s) - '/dev/null' saved [2/2],
,
INFO: Add CRON_ABORT="0 6 * * *" to cancel outstanding sync at 6am,
INFO: Starting crond ...,
INFO: crond started,
==> /tmp/crond.log <==,
,
==> /tmp/sync.log <==,
```

Log file:
```
2020/06/09 17:59:22 INFO  : Starting bandwidth limiter at 200kBytes/s
2020/06/09 17:59:26 INFO  : TEST DIR.txt: Copied (new)
2020/06/09 17:59:26 INFO  :
Transferred:             0 / 0 Bytes, -, 0 Bytes/s, ETA -
Transferred:            1 / 1, 100%
Elapsed time:         2.6s
```

**Comment by l4t3b0** [comment]
Hi pfidr34

I must be overlooking something, but i dont see the change reverted. I mean, it is reverted back to the state on 9th of june, but the working version is the one on 28 of may.

This snippet is from github master branch:
```
    if [ ! -z "$OUTPUT_LOG" ]
    then
      d=$(date +%Y_%m_%d-%H_%M_%S)
      LOG_FILE="/logs/$d.txt"
      echo "INFO: Log file output to $LOG_FILE"
      echo "INFO: Starting rclone $RCLONE_CMD $SYNC_SRC $SYNC_DEST $RCLONE_OPTS $SYNC_OPTS_ALL --log-file=${LOG_FILE}"
      set +e
      eval "rclone $RCLONE_CMD $SYNC_SRC $SYNC_DEST $RCLONE_OPTS $SYNC_OPTS_ALL --log-file=${LOG_FILE}"
      export RETURN_CODE=$?
      set -e
    else
      echo "INFO: Starting rclone $RCLONE_CMD $SYNC_SRC $SYNC_DEST $RCLONE_OPTS $SYNC_OPTS_ALL"
      set +e
      eval "rclone $RCLONE_CMD $SYNC_SRC $SYNC_DEST $RCLONE_OPTS $SYNC_OPTS_ALL"
      set -e
      export RETURN_CODE=$?
    fi
```
I dont see any chance to work with whitespaces if the src and dst are handled like the code above.

**Comment by l4t3b0** [comment]
May i ask what was the reason to put the whole rclone command in an eval?
Im very curious and id like to think about the problem also to keep the whitespaces work and achieve your goal too.

**Comment by pfidr34** [comment]
Change was to handle `--bwlimit "06:00,200 00:00,off"`

I'm not sure why white space wouldn't be working for you. I have proved it works for `SYNC_SRC` and `SYNC_DEST`

Do you have the paths wrapped in single ticks?

**Comment by l4t3b0** [comment]
hi pfidr34

Can you please verify my modification in the pull request please?

**Comment by pfidr34** [comment]
@l4t3b0 there are no changes needed. This is working as expected if you are using single quotes around your file paths.

The use of single quotes is described in the instructions for `SYNC_SRC` and `SYNC_DEST`

Docker Compose (notice spaces in `SYNC_SRC` and `SYNC_DEST`):
```
rclone_test:
    image: "rclone_test"
    container_name: "rclone_test"
    volumes:
      - ${USERDIR}/docker/rclone_test:/config
      - ${USERDIR}/docker/rclone_test/logs:/logs
      - /mnt/SSD_STRGE/TEMP/:/mnt/TEMP
    environment:
      - UID=1000
      - GID=1000
      - CRON=@daily
      - RCLONE_CMD=copy
      - SYNC_SRC='/mnt/TEMP/TEST DIR'
      - SYNC_DEST='GDrive:/1/TEST DEST'
      - SYNC_OPTS=--fast-list --transfers 10 -v --bwlimit "06:00,200 00:00,off"
      - RCLONE_DIR_CMD=lsf
      - RCLONE_DIR_CMD_DEPTH=1
      - OUTPUT_LOG=1
      - ROTATE_LOG=7
      - CHECK_URL=https://hc-ping.com/<snip>
      - FORCE_SYNC=1
      - TZ=${TZ}
```
Container log:
```
INFO: Running rclone v1.52.1
INFO: Cron shortcut @daily re-written to 0 0 * * *
INFO: Starting sync.sh pid 40 Wed Jun 24 13:14:27 CDT 2020
INFO: Sending start signal to healthchecks.io
--2020-06-24 13:14:27--  https://hc-ping.com/<snip>/start
Resolving hc-ping.com... 176.9.71.146, 178.63.26.145, 159.69.66.229, ...
Connecting to hc-ping.com|176.9.71.146|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 2 [text/plain]
Saving to: '/dev/null'
     0K                                                       100%  118K=0s
2020-06-24 13:14:28 (118 KB/s) - '/dev/null' saved [2/2]
INFO: Removing logs older than 7 day(s)...
INFO: Source directory is not empty and can be processed without clear loss of data
INFO: Log file output to /logs/2020_06_24-13_14_28.txt
INFO: Starting rclone copy '/mnt/TEMP/TEST DIR' 'GDrive:/1/TEST DEST' --config /config/rclone.conf --fast-list --transfers 10 -v --bwlimit "06:00,200 00:00,off" --log-file=/logs/2020_06_24-13_14_28.txt
INFO: Sending complete signal to healthchecks.io
--2020-06-24 13:14:39--  https://hc-ping.com/<snip>
Resolving hc-ping.com... 176.9.71.146, 178.63.26.145, 159.69.66.229, ...
Connecting to hc-ping.com|176.9.71.146|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 2 [text/plain]
Saving to: '/dev/null'
     0K                                                       100%  115K=0s
2020-06-24 13:14:39 (115 KB/s) - '/dev/null' saved [2/2]
INFO: Add CRON_ABORT="0 6 * * *" to cancel outstanding sync at 6am
INFO: Starting crond ...
INFO: crond started
==> /tmp/crond.log <==
==> /tmp/sync.log <==
```
Log file:
```
2020/06/24 13:14:28 INFO  : Starting bandwidth limiter at 200kBytes/s
2020/06/24 13:14:39 INFO  : TEST DIR.txt: Copied (new)
2020/06/24 13:14:39 INFO  :
Transferred:             0 / 0 Bytes, -, 0 Bytes/s, ETA -
Transferred:            1 / 1, 100%
Elapsed time:         4.0s
```

**Comment by l4t3b0** [comment]
Hi pfidr34

Sorry to disagree with your evaluation.
From my point of view your suggestion - to use single quotes as the part of the directory name - is a workaround and unfortunately it does not work in my case either.

**Why it is not working?**
Im using Unraid as the docker host environment. In docker im using UI and not executing the commands through CLI.
On the web ui i have a text field to enter the SYNC_SRC directory and Unraid is already using the single quotes.

My start up command - generated by Unraid - looks like this:
root@localhost:# /usr/local/emhttp/plugins/dynamix.docker.manager/scripts/docker run -d --name='rclone-alapitvany-private' --net='bridge' -e TZ="Europe/Budapest" -e HOST_OS="Unraid" -e 'SYNC_SRC'='webdav-alapitvany-private:/Megosztott dokumentumok/' -e 'SYNC_OPTS'='-v --exclude /Forms/**' -e 'SYNC_ONCE'='' -e 'CRON'='@weekly' -e 'CRON_ABORT'='' -e 'FORCE_SYNC'='1' -e 'SYNC_OPTS_EVAL'='' -e 'RCLONE_CMD'='sync' -e 'RCLONE_DIR_CMD'='lsf' -e 'RCLONE_DIR_CHECK_SKIP'='' -e 'CHECK_URL'='https://hc-ping.com/d34f5437-.......5b3f023cd' -e 'FAIL_URL'='' -e 'OUTPUT_LOG'='1' -e 'ROTATE_LOG'='120' -e 'SYNC_DEST'='/data' -v '/mnt/user/Cloud Backup/webdav-alapitvany-private/':'/data':'rw' -v '/mnt/user/appdata/rclone-webdav-alapitvany-private/':'/config':'rw' 'pfidr/rclone'

If i would use the single quote, than it simple does not work, because the value of the SYNC_SRC would start with 2 single quotes and it means empty directory.
So, basically with UI based docker environments you dont have the chance to do this kind of escaping :(

**Why i consider your solution a workaround?**
Without using single quotes the script gets the parameter value in a correct way. Why should anyone escape it if it is correct?
I believe the script should handle the parameter value differently.

Let me suggest you to check out my pull request code because i could fix it in the right way and your bwlimit parameter is working also.
If you accept my modification, than you and i could use your great project without unnecessary escaping parameters.

Thanks in advance

**Comment by l4t3b0** [comment]
Hi pfidr34

Could you share your opinion about about my evaluation, please

**Comment by pfidr34** [comment]
Saw you pushed your own big commit. Congrats!
